syntax = "proto3";

package scalekit.v1.commons;

import "buf/validate/validate.proto";
import "google/api/annotations.proto";
import "google/api/field_behavior.proto";
import "google/protobuf/any.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

option go_package = "github.com/scalekit-inc/scalekit/pkg/grpc/commons";

enum RegionCode {
  /*
     If the region code is unspecified the selected region will automatically be decided based on the origin of the call.
  */
  REGION_CODE_UNSPECIFIED = 0;
  US = 1;
  EU = 2;
}

enum EnvironmentType {
  ENVIRONMENT_TYPE_UNSPECIFIED = 0;
  PRD = 1;
  DEV = 2;
}

message OrganizationMembership {
  string id = 1;
  UserStatus membership_status = 2;
  MembershipRole role = 3;
  optional string name = 4;
  IdentityProviderType primary_identity_provider = 5;
}

enum UserStatus {
  USER_STATUS_UNSPECIFIED = 0;
  ACTIVE = 1;
  INACTIVE = 2;
}

enum MembershipRole {
  MEMBERSHIP_ROLE_UNSPECIFIED = 0;
  ADMIN = 1;
  USER = 2;
}

message UserProfile {
  string id = 1 [(google.api.field_behavior) = OUTPUT_ONLY];
  string first_name = 2 [(buf.validate.field).string = {max_len: 200}];
  string last_name = 3 [(buf.validate.field).string = {max_len: 200}];
  string name = 4;
  string locale = 5;
  bool email_verified = 6 [(google.api.field_behavior) = OUTPUT_ONLY];
  map<string, string> metadata = 7 [(buf.validate.field).map = {
    max_pairs: 30
    keys: {
      string: {
        min_len: 3
        max_len: 25
      }
    }
    values: {
      string: {
        min_len: 1
        max_len: 256
      }
    }
  }];
  map<string, string> custom_attributes = 8 [(buf.validate.field).map = {
    max_pairs: 100
    keys: {
      string: {
        min_len: 3
        max_len: 25
      }
    }
    values: {
      string: {
        min_len: 1
        max_len: 256
      }
    }
  }];
}

enum IdentityProviderType {
  IDENTITY_PROVIDER_UNSPECIFIED = 0;
  OKTA = 1;
  GOOGLE = 2;
  MICROSOFT_AD = 3;
  AUTH0 = 4;
  ONELOGIN = 5;
  PING_IDENTITY = 6;
  JUMPCLOUD = 7;
  CUSTOM = 8;
  GITHUB = 9;
  GITLAB = 10;
  LINKEDIN = 11;
  SALESFORCE = 12;
  MICROSOFT = 13;
  IDP_SIMULATOR = 14;
  SCALEKIT = 15;
}
